### ISSUE TITLE
Does not mark 'empty object' as decoded

### ISSUE THREAD

**Issue by michaeljones** [zwilias/json-decode-exploration#3]
I have a situation where I am applying a decoder to a field as below:

```elm
decode ActivityDisplay
    |> optional "color" (maybe string) Nothing
    |> optional "logo" (maybe string) Nothing

decode Activity
    |> required "id" int
    |> required "name" slugExplorationDecoder
    |> hardcoded type_
    |> required "display" displayDecoder
```

And it seems that if the Json has `"display": {}` then I get a warning saying that the `{}` is an `UnusedValue`. I'm trying to figure out a work around but I suspect that it isn't ideal behaviour from the library? I've attempted to decode some of it so I have 'used' it even if there was nothing there?

**Comment by michaeljones** [comment]
This is my approach for resolving it in my code. I can attempt a change in the library if you'd like though I suspect I'd get a little lost :) 

```elm
displayDecoder =
    let
        normalDecoder =
            decode ActivityDisplay
                |> optional "color" (maybe string) Nothing
                |> optional "logo" (maybe string) Nothing
    in
    normalDecoder
        |> Decode.andThen
            (\({ color, logo } as display) ->
                case ( color, logo ) of
                    ( Nothing, Nothing ) ->
                        Decode.keyValuePairs Decode.value
                            |> Decode.andThen
                                (\list ->
                                    if List.isEmpty list then
                                        Decode.succeed display
                                    else
                                        Decode.fail "Unexpected fields in display object"
                                )

                    _ ->
                        Decode.succeed display
            )
```

**Comment by zwilias** [comment]
Fix released in 4.2.1

Note that the fix for this changes error behaviour too. I explained this a little more in the README, the docs for `optional`, and encoded what I consider the correct behaviour in [the tests](https://github.com/zwilias/json-decode-exploration/blob/4.2.1/tests/PipelineTests.elm#L28-L62)

Thanks!

**Comment by michaeljones** [comment]
Thank for the work but I fear there might be an issue! #4
